3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/MailgunAdapter.scala

```diff
@@ -0,0 +1,214 @@
+    "unsubscribed" -> SchemaKey("com.mailgun", "recipient_unsubscribed", "jsonschema", "1-0-0").toSchemaUri
+  )
+
+  /**
+   * Converts a CollectorPayload instance into raw events.
+   *
+   * A Mailgun Tracking payload contains one single event
+   * in the body of the payload, stored within a HTTP encoded
+   * string.
+   *
+   * @param payload The CollectorPayload containing one or more
+   *        raw events as collected by a Snowplow collector
+   * @param resolver (implicit) The Iglu resolver used for
+   *        schema lookup and validation. Not used
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+def toRawEvents(payload: CollectorPayload)(implicit resolver: Resolver): ValidatedRawEvents = {
+    (payload.body, payload.contentType) match {
+      case (None, _)                                                => s"Request body is empty: no ${VendorName} events to process".failNel
+      case (_, None)                                                => s"Request body provided but content type empty, expected ${ContentTypesStr} for ${VendorName}".failNel
+      case (_, Some(ct)) if !ContentTypes.exists(ct.startsWith(_))  => s"Content type of ${ct} provided, expected ${ContentTypesStr} for ${VendorName}".failNel
+      case (Some(body), _) if (body.isEmpty)                        => s"${VendorName} event body is empty: nothing to process".failNel
+      case (Some(body), Some(ct))                                   => {
+        val params = toMap(payload.querystring)
+        try {
+          val bodyMap = getBoundary(ct)
+            .map(parseMultipartForm(body, _))
+            .getOrElse(toMap(URLEncodedUtils.parse(URI.create("http://localhost/?" + body), "UTF-8").toList))
+          bodyMap.get("event") match {
+            case None            => s"No ${VendorName} event parameter provided: cannot determine event type".failNel
+            case Some(eventType) => (payloadBodyToEvent(bodyMap)) match {
+              case (Failure(err)) => err.failNel
+              case (Success(event)) => {
+                lookupSchema(eventType.some, VendorName, EventSchemaMap) match {
+                  case Failure(str)  => str.fail
+                  case Success(schemaUri) => {
+                    NonEmptyList(RawEvent(
+                      api         = payload.api,
+                      parameters  = toUnstructEventParams(TrackerVersion, params, schemaUri, cleanupJsonEventValues(
+                        mutateMailgunEvent(event),
+                        ("event", eventType).some, "timestamp"), "srv"),
+                      contentType = payload.contentType,
+                      source      = payload.source,
+                      context     = payload.context
+                    )).success
+                  }
+                }
+              }
+            }
+          }
```

Comment:
can't we refactor those nested matches as:

```scala
            // ...
            case Some(eventType) =>
              for {
                event <- payloadBodyToEvent(bodyMap).leftMap(_.wrapNel)
                schemaUri <- lookupSchema(eventType.some, VendorName, EventSchemaMap)
              } yield NonEmptyList(RawEvent(
                api         = payload.api,
                parameters  = toUnstructEventParams(TrackerVersion, params, schemaUri, cleanupJsonEventValues(
                  mutateMailgunEvent(event),
                  ("event", eventType).some, "timestamp"), "srv"),
                contentType = payload.contentType,
                source      = payload.source,
                context     = payload.context
              ))
```

this will be way more readable imo